src/run_autodiscovery.py

```diff
@@ -0,0 +1,237 @@
+
+    dt_url = f"{dynatrace_url.rstrip('/')}/api/v2/settings/objects"
+    response = await dt_session.post(
+        url=dt_url,
+        headers={
+            "Authorization": f"Api-Token {dynatrace_access_key}",
+            "Content-Type": "application/json"
+        },
+        data=data_input,
+        verify_ssl=get_should_require_valid_certificate()
+    )
+
+    if response.status != 200:
+        print("Error During sending metadata meterics")
+        return False
+    else:
+        return True
+
+
+async def get_metric_descriptors(gcp_session,token):
+    headers = {
+    "Accept": "application/json",
+    "Authorization": f"Bearer {token}"}
+    url = f"https://monitoring.googleapis.com/v3/projects/{projectID}/metricDescriptors"
+    resp = await gcp_session.request('GET', url=url, headers=headers)
+    page = await resp.json()
+
+    metric_yaml = [
+        {
+            "value": x.get("type"),
+            "key":  "cloud.gcp." + metric_name_parse(x.get("type"),type_metrickind_caster(x.get("metricKind"),x.get("valueType"))),
+            "displayName":  x.get("displayName"),
+            "name": x.get("displayName"),
+            "description": x.get("description"),
+            "type": type_metrickind_caster(x.get("metricKind"),x.get("valueType")),
+            "gcpOptions":
+                {
+                    "ingestDelay":  int(x.get("metadata",{}).get("ingestDelay","60s")[:-1]),
+                    "samplePeriod": int(x.get("metadata",{}).get("samplePeriod","60s")[:-1]),
+                    "valueType":    x.get("valueType"),
+                    "metricKind":   x.get("metricKind"),
+                    "unit":         unit_caster(x.get("unit"))
+            },
+            "dimensions": [ 
+                {
+                "key":      dimension.get("key"), 
+                "value":    "label:metric.labels." + dimension.get("key"),
+                } for dimension in x.get("labels") or []],
+            "monitored_resources_types": x.get("monitoredResourceTypes",[])
+        } for x in page["metricDescriptors"]]
+    
+    
+
+
+
+    metric_list = [(Metric(**x), x.get("monitored_resources_types"),x) for x in metric_yaml if x.get("gcpOptions").get("valueType").upper() != "STRING" ]
+    discovered_resource_type_list = list(filter(lambda x: discovered_resource_type in x[1], metric_list))
+    return discovered_resource_type_list
+
+
+async def enrich_services_autodiscovery(gcp_services_list,gcp_session,dt_session,token):
+    print("Start AutoDiscovery")
+
+
+    bucket_gcp_services =  list(filter(lambda x: discovered_resource_type in x.name, gcp_services_list[0]))
+    existing_metric_list = [metric for service in bucket_gcp_services for metric in service.metrics]
+    all_metrics_list = await get_metric_descriptors(gcp_session,token)
+
+
+    missing_metrics_list = [metric for metric in all_metrics_list if metric[0].google_metric not in [existing_metric.google_metric for existing_metric in existing_metric_list]]
```

Comment:
This is kinda packed, maybe lets try to extract some additional variables, then it is easier to read:
gcp_keys_from_metric_list = XXX
And then list comprehension, at least you will not nest it, there is no good reason to nest here